File: ts/src/domain/strategy/RandomMoveStrategy/RandomMoveStrategy.ts
import { MoveStrategy } from '../MoveStrategy';

export class RandomMoveStrategy implements MoveStrategy {
    private static readonly MOVE_THRESHOLD = 4;
    private static readonly RANDOM_RANGE = 10;

    canMove(): boolean {
        const randomValue = Math.floor(Math.random() * RandomMoveStrategy.RANDOM_RANGE);
        return randomValue >= RandomMoveStrategy.MOVE_THRESHOLD;
    }
}

File: java/src/main/java/racing/strategy/RandomMoveStrategy.java
package racing.strategy;

import java.util.function.IntSupplier;

public class RandomMoveStrategy implements MoveStrategy {
    private static final int FORWARD_THRESHOLD = 4;
    private static final int RANDOM_BOUND = 10;
    private final IntSupplier randomProvider;

    public RandomMoveStrategy() {
        this(() -> (int) (Math.random() * RANDOM_BOUND));
    }

    public RandomMoveStrategy(IntSupplier randomProvider) {
        this.randomProvider = randomProvider;
    }

    @Override
    public boolean canMove() {
        return this.randomProvider.getAsInt() >= FORWARD_THRESHOLD;
    }
}
